src/Symfony/Component/Messenger/Transport/Serialization/Serializer.php

```diff
@@ -62,12 +69,12 @@ public static function create(): self
      */
     public function decode(array $encodedEnvelope): Envelope
     {
-        if (empty($encodedEnvelope['body']) || empty($encodedEnvelope['headers'])) {
-            throw new MessageDecodingFailedException('Encoded envelope should have at least a "body" and some "headers".');
-        }
-
-        if (empty($encodedEnvelope['headers']['type'])) {
-            throw new MessageDecodingFailedException('Encoded envelope does not have a "type" header.');
+        $headersRequired = !isset($this->context[self::TYPE_RESOLVER]) || $this->context[self::TYPE_RESOLVER] instanceof HeaderTypeResolver;
```

Comment:
This looks weird to me - we're hardcoding a special case for `HeaderTypeResolver`. Is this the exception needed anymore? Or could we fail *later* with some message about needing body/headers or a type resolver?

Reply:
It was to stay close to the original behavior. But a change in the exception message can probably not be considered as a BC break :)